internal/planner/planner.go

```diff
@@ -2415,13 +2415,23 @@ outer:
 						nextNodes = append(nextNodes, node)
 					}
 				}
+				if len(node.Rules()) > 0 {
+					refs := refsOfRules(node.Rules())
+					p.debugf("no optimization of %s: node with rules (%v)", ref, refs)
+					return dont()
+				}
 			}
 		case ast.String:
 			// take all children that either match or have a var key // TODO(sr): Where's the code for the second part, having a var key?
 			for _, node := range nodes {
 				if node := node.Get(r); node != nil {
 					nextNodes = append(nextNodes, node)
 				}
+				if len(node.Rules()) > 0 {
+					refs := refsOfRules(node.Rules())
+					p.debugf("no optimization of %s: node with rules (%v)", ref, refs)
+					return dont()
+				}
```

Comment:
We're moving a part of the "unbalanced rule trie" check up into the loop. Previously, we hadn't checked node.Rules() at all, so the rule of `p.allow[action][resource]` was dropped when we passed "action" -- the rest of the rule isn't part of the ruletrie, it's part of the rule body.

Now, we'll check when combing through the rule trie if some such ref rule is present, or any shorter-than expected rule.

Technically, even a rule like
```
p.allow := x if {
  x := { "list": { "fruit": true }}
}
```
_cannot_ be neglected here.